### DISCUSSION TITLE
How to call `cmp.complete()` inside a telescope action?

### DISCUSSION THREAD

**Discussion by Sleepful** [hrsh7th/nvim-cmp#1629] [Q&A] [answered]
I want to call `cmp.complete()` after a telescope.nvim search

```lua
local telescope_cmp = function(opts)
  opts = opts or {}
  pickers
    .new(opts, {
      prompt_title = "Auto CMP",
      finder = finders.new_table({
        results = entries,
        entry_maker = function(entry)
          return ...
        end,
      }),
      sorter = conf.generic_sorter(opts),
      attach_mappings = function(prompt_bufnr, map)
        actions.select_default:replace(function()
          actions.close(prompt_bufnr)
          vim.schedule(function()
            vim.cmd("startinsert!")
            cmp.complete() -------- does not work ------- :(
          end)

        end)
        return true
      end,
    })
    :find()
end

-- cmp.complete() -- over here it DOES work, so IDK why the other callsite does not
telescope_cmp()

```

**Comment by Shougo** [comment]
Why the feature is needed?

**Reply by Sleepful** [reply]
I am using telescope to pick an option from the CMP entries. But I am having difficulty expanding the entry. Ideally I use CMP `confirm` to expand the entry. But for some reason I cannot re_open CMP after I am done with Telescope.

**Reply by Sleepful** [reply]
I am following https://github.com/nvim-telescope/telescope.nvim/blob/master/developers.md for the telescope code. Straightforward.

**Reply by Sleepful** [reply]
I guess it seems like telescope is breaking _something_ required by CMP to properly activate, no idea what it could be.

**Reply by Shougo** [reply]
Hm...  It seems difficult.
nvim-cmp doesn't know what item is inserted.

**Reply by Sleepful** [reply]
it would be enough for me to work if the nvim cmp can _open_ (through `cmp.complete()`) after the telescope closes.
do you know why it cannot open?

**Reply by Sleepful** [reply]
do you know how I could make the cmp `complete()` menu open after the telescope function ends?

**Reply by Shougo** [reply]
Ah, I think nvim-cmp doesn't work in normal mode.
startinsert doesn't change the mode immediate.

**Reply by Sleepful** [reply]
oh hey that is probably it, I need to figure out how to change the mode. Thank you.

**Reply by Sleepful** [reply]
Yep, this fails just the same:
```lua
              vim.cmd([[stopinsert]])
              vim.schedule(function()
                vim.cmd([[startinsert]])
                require("cmp").complete()
              end)
```

**Reply by Sleepful** [reply]
well at least I am happy it has nothing to do with telescope

**Reply by Sleepful** [reply]
Do I need to use `feedkeys`?

**Reply by yamada-wacul** [reply]
```diff
-            vim.cmd("startinsert!")
-            cmp.complete()
+            vim.api.nvim_create_autocmd("ModeChanged", {
+                pattern = "*:i",
+                once = true,
+                callback = function()
+                    cmp.complete()
+                end,
+            })
+            vim.cmd("startinsert!")
```

@Sleepful Does this work well?
Now I don't use Telescope anymore, so sorry if I'm wrong.

**Reply by hrsh7th** [reply, marked as answer]
```lua
        actions.select_default:replace(function()
          actions.close(prompt_bufnr)
          vim.schedule(function()
            vim.cmd("startinsert!")
            cmp.complete() -------- does not work ------- :(
          end)
        end)
```
↓↓↓
```lua
        actions.select_default:replace(function()
          actions.close(prompt_bufnr)
          vim.schedule(function()
            vim.fn.feedkeys(vim.api.nvim_replace_termcodes('i<Cmd>lua require("cmp").complete()<CR>', true, true, true))
          end)

        end)
```

**Reply by Sleepful** [reply]
Oh! That works! Amazing thank you!

Minimal example of solution without telescope:

```lua
vim.cmd([[stopinsert]])
vim.schedule(function()
  vim.fn.feedkeys(
    vim.api.nvim_replace_termcodes('i<Cmd>lua require("cmp").complete()<CR>', true, true, true)
  )
end)
```

Why is `feedkeys` necessary?

Seems like a limitation with Neovim.

**Reply by Shougo** [reply]
Because feedkeys is executed immediately.

**Reply by Sleepful** [reply]
@Shougo sorry I do not think I understand the meaning of “immediately” in this context.

Is the lua function not executed _immediately_?

Why is it necessary to execute the function _immediately_? Even if it does not execute _immediately_ I would still expect it to execute properly.

**Reply by Shougo** [reply]
startinsert doesn't change the state immediately.
It schedules the sate change.
Lua function is executed immediately.

**Reply by Sleepful** [reply]
@Shougo weird! Thanks

**Reply by Sleepful** [reply]
so using `feedkeys` solves this problem because `feedkeys` executes the lua function with the same delay as `i`?

**Reply by Shougo** [reply]
`feedkeys()` is not Lua function.
It is just Vim/neovim feature.

**Reply by Sleepful** [reply]
oh yeah I meant, the function in the text string. The lua function in the text string. This part:
```
'i<Cmd>lua require("cmp").complete()<CR>'
```
my understanding is that `feedkeys` will allow `complete()` to delay just as much as the insert `i`. That way it will work.
Not sure if this is correct.

**Reply by Sleepful** [reply]
I am trying to improve my understanding of neovim :)

**Reply by Shougo** [reply]
The `complete()` is executed as key mapping.
so it is executed immediately.
